### PR TITLE
Add tests against local backend

### PR DESCRIPTION
**Pull request by sshader**
Instructions are to run a local backend, deploy to it, set the `IS_TEST` environment variable, and then tests are runnable.

A few interesting bits:
* I made `testingQuery` wrappers that error if the `IS_TEST` env var isn't set. Right now they're used for clearing data between tests, but could potentially be used for setting up data or even running logic + assertions (like our isolate tests for built-ins)
* I used the auth impersonation to run auth-ed queries -- I wasn't really sure of a better way to do this since I don't think Auth0 makes it all that easy to grab a token from a test even if you set up a separate app with test credentials

### CODE REVIEW

File: convex/testing/clearAll.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { testingMutation } from "./wrappers";
+import schema from "../schema";
+
+export default testingMutation(async ({ db }) => {
+    for (const table of Object.keys(schema.tables)) {
+      const docs = await db.query(table as any).collect();
```
Comment (xixixao) [inline convex/testing/clearAll.ts:6]: would be great to get typechecking working here (not important though)

File: jest.config.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
+module.exports = {
+    preset: "ts-jest/presets/js-with-ts",
+    testEnvironment: "node",
+    testTimeout: 10000,
+  
+    // Only run one suite at a time because all of our tests are running against
+    // the same backend and we don't want to leak state.
+    maxWorkers: 1,
```
Comment (xixixao) [inline jest.config.js:9]: One downside of not doing this in JS

File: tests.test.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    return {
+      ...args,
+      subject,
+      issuer
+    }
+  }
+}
+
+function setAuth(client: ConvexReactClient, identity: Omit<UserIdentity, "tokenIdentifier">) {
+  const ADMIN_KEY = '0135d8598650f8f5cb0f30c34ec2e2bb62793bc28717c8eb6fb577996d50be5f4281b59181095065c5d0f86a2c31ddbe9b597ec62b47ded69782cd';
```
Comment (xixixao) [inline tests.test.ts:25]: How do people get this, and how should they configure it? Env var (node.js one)?

File: tests.test.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  });
+
+  afterEach(async () => {
+    await client.mutation(api.testing.clearAll.default, {});
+    await client.close()
+  });
+
+  test("two players can join game", async () => {
+    const sarahIdentity = identityFactory.newIdentity({ name: "Sarah" })
+    setAuth(client, sarahIdentity)
```
Comment (xixixao) [inline tests.test.ts:47]: This could be simplified maybe?

File: tests.test.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    await client.mutation(api.games.move, { gameId, from: "c2", to: "c3" })
+    game = await client.query(api.games.get, { id: gameId })
+    expect(game.pgn).toEqual("1. c3")
+    
+    // Invalid move -- out of turn
+    expect(() => client.mutation(api.games.move, { gameId, from: "d2", to: "d3" })).rejects.toThrow(ConvexError)
+    game = await client.query(api.games.get, { id: gameId })
+    expect(game.pgn).toEqual("1. c3")
+  });
+});
```
Comment (xixixao) [inline tests.test.ts:68]: Stylistically I wouldn't want to repeat `client.` constantly

Comment (xixixao) [inline tests.test.ts:68]: Can you add an example that uses `testingMutation` to scaffold some data (for the cases where scaffolding through public methods is cumbersome)

File: convex/testing/wrappers.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+export  const testingAction = customAction(action, {
+    args: {},
+    input: async (_ctx, _args) => {
+      if (process.env.IS_TEST === undefined) {
+          throw new Error("Calling a test only function in an unexpected environment")
+      }
+      return { ctx: {}, args: {} };
+    },
+  });
```
Comment (xixixao) [inline convex/testing/wrappers.ts:32]: Probably not needed

File: tests.test.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  (client as any).setAdminAuth(ADMIN_KEY, identity)
+}
+
+
+describe("games", () => {
+  let client: ConvexReactClient;
+  let identityFactory: IdentityFactory
+  beforeEach(() => {
+    // @ts-ignore
+    client = new ConvexReactClient("http://127.0.0.1:8000", { webSocketConstructor: WebSocket});
```
[the remark this answers is not in the document]
Comment (xixixao) [reply tests.test.ts:36]: Do we have to restart the client between each test?

File: convex/testing/clearAll.ts (hunk 1 above)
Comment (sshader) [reply convex/testing/clearAll.ts:6]: It's the `Object.keys` that gives us `string` but maybe there's a different way to iterate through

File: tests.test.ts (hunk 3 above)
Comment (sshader) [reply tests.test.ts:25]: I think the plan is just to bake this into all local backends (I copied it from the Justfile from that repo)

File: tests.test.ts (hunk 5 above)
Comment (sshader) [reply tests.test.ts:68]: Any suggestion beyond just shortening it (my default abbreviation is `t` for "tests")

File: tests.test.ts (hunk 4 above)
Comment (sshader) [reply tests.test.ts:47]: Simplified as in combining this with the line above? Or closing over the client / attaching directly to the client?

I thought about combining these and having it return the generated identity but I kind of liked being super explicit about creating new identities (while also not having to deal with creating unique subjects)

File: tests.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (sshader) [reply tests.test.ts:36]: This is what we're doing in js integration tests. Idk if there's a lighter-weight thing we could do, but we definitely do want to reset state.

Also I realized the one test I was looking at that didn't pass in `ws` was in the jest-dom environment, and the other ones all do (with a type cast hidden in there)

File: tests.test.ts (hunk 5 above)
Comment (sshader) [reply tests.test.ts:136]: I tried a builder pattern which made it a little nicer I think ¯\\\_(ツ)\_/¯

File: tests.test.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        (this.client as any).setAdminAuth(this._adminKey, identity);
+        return this.client.action(functionReference, args).finally(() => {
+          this.client.client.clearAuth()
+        })
+      },
+      query: (functionReference, args) => {
+        (this.client as any).setAdminAuth(this._adminKey, identity);
+        return this.client.query(functionReference, args).finally(() => {
+          this.client.client.clearAuth()
+        })
```
Comment (xixixao) [inline tests.test.ts:46]: What about HTTPClient, so it's stateless? This way I cannot test errors

**Pull request closed without merging**